### PR TITLE
Reduce memory usage of incremental compiler

### PR DESCRIPTION
**Pull request by oehme**
This reduces the long term heap usage of our largeJavaMultiProject by about 25%. The main changes are:

- using immutable data structures in the caches instead of inefficient HashMap/HashSet
- de-duplicating empty constant sets
- snapshotting class directories on the classpath the same way we snapshot jars instead of duplicating their analysis in every compile task

### CODE REVIEW

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/CachingClasspathEntrySnapshotter.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    public CachingClasspathEntrySnapshotter(StreamHasher streamHasher, FileSystemSnapshotter fileSystemSnapshotter, ClassDependenciesAnalyzer analyzer, ClasspathEntrySnapshotCache cache) {
+        this.snapshotter = new DefaultClasspathEntrySnapshotter(streamHasher, analyzer);
+        this.fileSystemSnapshotter = fileSystemSnapshotter;
+        this.cache = cache;
+    }
+
+    @Override
+    public ClasspathEntrySnapshot createSnapshot(final ClasspathEntry classpathEntry) {
+        final HashCode hash = getHash(classpathEntry);
```
Comment (marcphilipp) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/CachingClasspathEntrySnapshotter.java]: Do we need to compute the hash eagerly or can we do it in the `Factory`?

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/ClasspathSnapshotMaker.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        maybeInitialize(classpath); //clients may or may not have already created classpath snapshot
+        return classpathSnapshot;
+    }
+
+    private void maybeInitialize(Iterable<File> classpath) {
+        if (classpathSnapshot != null) {
+            return;
+        }
+        Timer clock = Time.startTimer();
+        Iterable<ClasspathEntry> entriesArchives = classpathEntryConverter.asClasspathEntries(classpath);
```
Comment (marcphilipp) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/ClasspathSnapshotMaker.java]: `entries`?

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/ClasspathSnapshotMaker.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+        classpathSnapshot = classpathSnapshotFactory.createSnapshot(entriesArchives);
+        int duplicatesCount = classpathSnapshot.getData().getDuplicateClasses().size();
+        String duplicateClassesMessage = duplicatesCount == 0? "" : ". " + duplicatesCount + " duplicate classes found in classpath (see all with --debug)";
```
Comment (marcphilipp) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/ClasspathSnapshotMaker.java]: Nit: missing space before `?`

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/CachingClasspathEntrySnapshotter.java (hunk 1 above)
Comment (oehme) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/CachingClasspathEntrySnapshotter.java]: I'm gonna clean this up in the next step, the whole package is pretty messy.

**Pull request merged**